quodlibet/quodlibet/ext/events/waveformseekbar.py

```diff
@@ -0,0 +1,277 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 0x1777
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 as
+# published by the Free Software Foundation
+
+from gi.repository import Gtk, Gdk, Gst
+
+from quodlibet import _, app, config
+from quodlibet.plugins.events import EventPlugin
+from quodlibet.qltk import Align
+from quodlibet.qltk.seekbutton import TimeLabel
+from quodlibet.qltk.tracker import TimeTracker
+from quodlibet.util import connect_destroy
+
+
+class WaveformSeekBar(Gtk.Box):
+    """ A widget containing labels and the seekbar."""
+
+    def __init__(self, player, library):
+        super(WaveformSeekBar, self).__init__()
+
+        self._player = player
+
+        self._elapsed_label = TimeLabel()
+        self._remaining_label = TimeLabel()
+        self._waveform_scale = WaveformScale()
+
+        self.pack_start(Align(self._elapsed_label, border=6), False, True, 0)
+        self.pack_start(self._waveform_scale, True, True, 0)
+        self.pack_start(Align(self._remaining_label, border=6), False, True, 0)
+
+        for child in self.get_children():
+            child.show_all()
+
+        self._tracker = TimeTracker(player)
+        self._tracker.connect('tick', self._on_tick, player)
+
+        connect_destroy(player, 'seek', self._on_player_seek)
+        connect_destroy(player, 'song-started', self._on_song_started)
+        connect_destroy(player, 'notify::seekable', self._on_seekable_changed)
+        connect_destroy(library, 'changed', self._on_song_changed, player)
+
+        self.connect('destroy', self._on_destroy)
+        self._update(player)
+        self._tracker.tick()
+
+    def _create_waveform(self, file):
+        # Close any existing pipelines to avoid warnings
+        if hasattr(self, "_pipeline") and self._pipeline:
+            self._pipeline.set_state(Gst.State.NULL)
+
+        command_template = """
+        filesrc location="{}"
+        ! decodebin
+        ! level name=audiolevel interval=100000000 post-messages=true
+        ! fakesink sync=false"""
+        command = command_template.format(file)
+        pipeline = Gst.parse_launch(command)
+
+        bus = pipeline.get_bus()
+        self._bus_id = bus.connect("message", self._on_bus_message)
+        bus.add_signal_watch()
+
+        pipeline.set_state(Gst.State.PLAYING)
+
+        self._pipeline = pipeline
+        self._rms_vals = []
+
+    def _on_bus_message(self, bus, message):
+        if message.type == Gst.MessageType.ERROR:
+            error, debug = message.parse_error()
+            print("Error received from element {name}: {error}".format(
```

Comment:
Please use `quodlibet.print_d` instead of print() everywhere. It adds logging and works better under Windows.